components/org.wso2.carbon.identity.cors.valve/src/main/java/org/wso2/carbon/identity/cors/valve/internal/handler/CORSRequestHandler.java

```diff
@@ -0,0 +1,269 @@
+        }
+
+        // Append response headers.
+        if (config.isSupportsCredentials()) {
+            response.addHeader(Header.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+
+            // The string "*" cannot be used for a resource that supports credentials.
+            response.addHeader(Header.ACCESS_CONTROL_ALLOW_ORIGIN, requestOrigin.toString());
+
+            // See https://bitbucket.org/thetransactioncompany/cors-filter/issue/16/
+            response.addHeader(Header.VARY, "Origin");
+        } else {
+            if (config.isAllowAnyOrigin()) {
+                response.addHeader(Header.ACCESS_CONTROL_ALLOW_ORIGIN, "*");
+            } else {
+                response.addHeader(Header.ACCESS_CONTROL_ALLOW_ORIGIN, requestOrigin.toString());
+
+                // See https://bitbucket.org/thetransactioncompany/cors-filter/issue/16/
+                response.addHeader(Header.VARY, "Origin");
+            }
+        }
+
+        if (!config.getExposedHeaders().isEmpty()) {
+            String exposedHeaders = HeaderUtils.serialize(config.getExposedHeaders(), ", ");
+            response.addHeader(Header.ACCESS_CONTROL_EXPOSE_HEADERS, exposedHeaders);
+        }
+    }
+
+    /**
+     * Handles a preflight CORS request.
+     *
+     * @param request  The HTTP request.
+     * @param response The HTTP response.
+     * @throws CORSException                  If the request is invalid or denied.
+     * @throws CORSManagementServiceException
+     */
+    public void handlePreflightRequest(HttpServletRequest request, HttpServletResponse response) throws
+            CORSManagementServiceException, CORSException {
+
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        CORSConfiguration config = getCORSManager().getCORSConfiguration(tenantDomain);
+
+        // Check origin against allow list
+        Origin requestOrigin = new Origin(request.getHeader(Header.ORIGIN));
+        if (!isAllowedOrigin(tenantDomain, requestOrigin)) {
+            throw new CORSException(ErrorMessages.ERROR_CODE_ORIGIN_DENIED);
+        }
+
+        // Parse requested method
+        // Note: method checking must be done after header parsing, see CORS spec.
+        String requestMethodHeader = request.getHeader(Header.ACCESS_CONTROL_REQUEST_METHOD);
+        if (requestMethodHeader == null) {
+            throw new CORSException(ErrorMessages.ERROR_CODE_MISSING_ACCESS_CONTROL_REQUEST_METHOD_HEADER);
+        }
+        final String requestedMethod = requestMethodHeader.toUpperCase();
+
+        // Parse the requested author (custom) headers
+        final String rawRequestHeadersString = request.getHeader(Header.ACCESS_CONTROL_REQUEST_HEADERS);
+        final String[] requestHeaderValues = HeaderUtils.parseMultipleHeaderValues(rawRequestHeadersString);
+
+        final String[] requestHeaders = new String[requestHeaderValues.length];
+
+        for (int i = 0; i < requestHeaders.length; i++) {
```

Comment:
cant we use a for-each loop here

Reply:
We are populating the elements in `requestHeaders`. Therefore it is not possible to use a for-each loop.